File: LISTAS/LISTA CICRULAR DOBLEMENTE ENLAZADA/LISTACIRCULARDOBLEMENTEENLAZADA.py
class NodoCircularDoble:
    def __init__(self, dato):
        self.dato = dato
        self.siguiente = None
        self.anterior = None

class ListaCircularDoble:
    def __init__(self):
        self.ultimo = None
    
    def insertar_vacio(self, dato):
        nuevo = NodoCircularDoble(dato)
        nuevo.siguiente = nuevo
        nuevo.anterior = nuevo
        self.ultimo = nuevo
    
    def insertar_inicio(self, dato):
        if not self.ultimo:
            self.insertar_vacio(dato)
            return
        
        nuevo = NodoCircularDoble(dato)
        nuevo.siguiente = self.ultimo.siguiente
        nuevo.anterior = self.ultimo
        self.ultimo.siguiente.anterior = nuevo
        self.ultimo.siguiente = nuevo
    
    def insertar_final(self, dato):
        if not self.ultimo:
            self.insertar_vacio(dato)
            return
        
        nuevo = NodoCircularDoble(dato)
        nuevo.siguiente = self.ultimo.siguiente
        nuevo.anterior = self.ultimo
        self.ultimo.siguiente.anterior = nuevo
        self.ultimo.siguiente = nuevo
        self.ultimo = nuevo
    
    def eliminar(self, dato):
        if not self.ultimo:
            return
        
        # Caso único elemento
        if self.ultimo.siguiente == self.ultimo and self.ultimo.dato == dato:
            self.ultimo = None
            return
        
        actual = self.ultimo.siguiente
        
        while True:
            if actual.dato == dato:
                actual.anterior.siguiente = actual.siguiente
                actual.siguiente.anterior = actual.anterior
                
                if actual == self.ultimo:
                    self.ultimo = actual.anterior
                return
            
            actual = actual.siguiente
            if actual == self.ultimo.siguiente:
                break
    
    def mostrar_adelante(self):
        if not self.ultimo:
            return []
        
        elementos = []
        actual = self.ultimo.siguiente
        
        while True:
            elementos.append(actual.dato)
            actual = actual.siguiente
            if actual == self.ultimo.siguiente:
                break
        
        return elementos
    
    def mostrar_atras(self):
        if not self.ultimo:
            return []
        
        elementos = []
        actual = self.ultimo
        
        while True:
            elementos.append(actual.dato)
            actual = actual.anterior
            if actual == self.ultimo:
                break
        
        return elementos

File: LISTAS/LISTA CICRULAR DOBLEMENTE ENLAZADA/LISTACIRCULARDOBLEMENTEENLAZADA.js
class NodoCircularDoble {
    constructor(dato) {
        this.dato = dato;
        this.siguiente = null;
        this.anterior = null;
    }
}

class ListaCircularDoble {
    constructor() {
        this.ultimo = null;
    }
    
    insertarVacio(dato) {
        const nuevo = new NodoCircularDoble(dato);
        nuevo.siguiente = nuevo;
        nuevo.anterior = nuevo;
        this.ultimo = nuevo;
    }
    
    insertarInicio(dato) {
        if (!this.ultimo) {
            this.insertarVacio(dato);
            return;
        }
        
        const nuevo = new NodoCircularDoble(dato);
        nuevo.siguiente = this.ultimo.siguiente;
        nuevo.anterior = this.ultimo;
        this.ultimo.siguiente.anterior = nuevo;
        this.ultimo.siguiente = nuevo;
    }
    
    insertarFinal(dato) {
        if (!this.ultimo) {
            this.insertarVacio(dato);
            return;
        }
        
        const nuevo = new NodoCircularDoble(dato);
        nuevo.siguiente = this.ultimo.siguiente;
        nuevo.anterior = this.ultimo;
        this.ultimo.siguiente.anterior = nuevo;
        this.ultimo.siguiente = nuevo;
        this.ultimo = nuevo;
    }
    
    eliminar(dato) {
        if (!this.ultimo) return;
        
        // Caso único elemento
        if (this.ultimo.siguiente === this.ultimo && this.ultimo.dato === dato) {
            this.ultimo = null;
            return;
        }
        
        let actual = this.ultimo.siguiente;
        
        do {
            if (actual.dato === dato) {
                actual.anterior.siguiente = actual.siguiente;
                actual.siguiente.anterior = actual.anterior;
                
                if (actual === this.ultimo) {
                    this.ultimo = actual.anterior;
                }
                return;
            }
            actual = actual.siguiente;
        } while (actual !== this.ultimo.siguiente);
    }
    
    mostrarAdelante() {
        if (!this.ultimo) return [];
        
        const elementos = [];
        let actual = this.ultimo.siguiente;
        
        do {
            elementos.push(actual.dato);
            actual = actual.siguiente;
        } while (actual !== this.ultimo.siguiente);
        
        return elementos;
    }
    
    mostrarAtras() {
        if (!this.ultimo) return [];
        
        const elementos = [];
        let actual = this.ultimo;
        
        do {
            elementos.push(actual.dato);
            actual = actual.anterior;
        } while (actual !== this.ultimo);
        
        return elementos;
    }
}